static/app/components/smartSearchBar/utils.tsx

```diff
@@ -317,3 +339,106 @@ export const shortcuts: Shortcut[] = [
     },
   },
 ];
+
+/**
+ * Groups tag keys based on the "." character in their key.
+ * For example, "device.arch" and "device.name" will be grouped together as children of "device", a non-interactive parent.
+ * The parent will become interactive if there exists a key "device".
+ */
+export const getTagItemsFromKeys = (
+  tagKeys: string[],
+  supportedTags: {
+    [key: string]: Tag;
+  },
+  getFieldDoc?: (key: string) => React.ReactNode
+) => {
+  return [...tagKeys]
+    .sort((a, b) => a.localeCompare(b))
+    .reduce((groups, key) => {
+      const keyWithColon = `${key}:`;
+      const sections = key.split('.');
+      const kind = supportedTags[key]?.kind;
+      const documentation = getFieldDoc?.(key) || '-';
+
+      const item: SearchItem = {
+        value: keyWithColon,
+        title: key,
+        documentation,
+        kind,
+      };
+
+      const lastGroup = groups.at(-1);
+
+      const [title] = sections;
+
+      if (kind !== FieldValueKind.FUNCTION && lastGroup) {
+        if (lastGroup.children && lastGroup.title === title) {
+          lastGroup.children.push(item);
+          return groups;
+        }
+
+        if (lastGroup.title && lastGroup.title.split('.')[0] === title) {
+          if (lastGroup.title === title) {
+            return [
+              ...groups.slice(0, -1),
+              {
+                title,
+                value: lastGroup.value,
+                documentation: lastGroup.documentation,
+                kind: lastGroup.kind,
+                children: [item],
+              },
+            ];
+          }
+
+          // Add a blank parent if the last group's full key is not the same as the title
+          return [
+            ...groups.slice(0, -1),
+            {
+              title,
+              value: null,
+              documentation: '-',
+              kind: lastGroup.kind,
+              children: [lastGroup, item],
+            },
+          ];
+        }
+      }
+
+      return [...groups, item];
+    }, [] as SearchItem[]);
+};
+
+/**
+ * Sets an item as active within a search group array.
+ * Comparison using the value of the item, assuming each item has a unique value.
+ */
+export const setSearchGroupItemActive = (
```

Comment:
this function is weird to me, it's weird to modify objects using .find

Reply:
not sure if a for loop makes more sense

Reply:
@scttcper Actually, do you think that instead of mutating `searchGroups` in place would it be better to just map it to a new `SearchGroup[]`?

Reply:
yeah that might be less of a mystery